core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1202,8 +1241,9 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
                     }
                   }
                 } else {
-                  if (!ufs.deleteDirectory(ufsUri, alluxio.underfs.options.DeleteOptions
-                      .defaults().setRecursive(true))) {
+                  if (!ufs.deleteDirectory(ufsUri, alluxio.underfs.options.DeleteOptions.defaults()
+                      .setRecursive(safeRecursiveUFSDeletes.containsKey(alluxioUriToDel)))) {
```

Comment:
It seems like we are selectively deleting certain directories recursively according to the safety check. With this approach, it could be the case that there are still a lot of UFS calls.

Instead, shouldn't we just do the initial check for the root of the delete, and then if the check succeeds, do a single UFS recursive delete on the UFS directory?